### PR TITLE
[rbgen 1/2]: preparation (don't squash)

### PR DESCRIPTION
**Pull request by winex**
okay, this i'd like to be reviewed, tested and merged into main branch. then i can continue on `rbgen`, because it needs almost all of this...

**release is not necessary** unless you want to provide rare crash fixes to current players (or who deletes chunks)
this ~~is~~ **was** mostly dev/chore PR with only ~~1 feature~~ 2 features:
- `map_gen_settings.starting_area` being supported for those who reserves 0,0 for bases, but it doesn't move `Spawn Initial City` thing yet
- while deferred tag placement is of no use right now, it shouldn't do anything wrong.
- **NEW:** check for nearby resources on industry placement and `Skip generation check: resources` option
- **NEW:** city population is shown in their names with `Show population` option

_p.s. don't know how you want your repository to look, but i would like these commits not to be squashed if possible, so we can look through history and revert/disable stuff for any reason. i didn't put that amount of information inside commit messages this time, but tried to keep them minimal and isolated as much as possible.
p.p.s. my previous (#265) commits had useful information, but squash-merge removed it completely. :( so i had to create [pr-1](/winex/factorio-tycoon/tree/pr-1) branch to keep that history_

hope this complies with your standards! :) and, of course, i always rebase my branches against latest commit, so things look nice and clean!

_**edits:** rebased, added 2nd feature_

### CODE REVIEW

File: chunk-charted-handler.lua
Comment (winex) [inline chunk-charted-handler.lua:1]: this one should be renamed to `chunk-handler.lua` as it has 3 handlers now, but i think renaming files should be done in **refactor commits** and isolated from code changes, so i kept that

**Comment by bahrmichael** [comment]
Heya! I should have time for more intensive testing on Saturday. Thank you so much for pushing this forward, sharing so much context, and pushing for high code quality. I really appreciate that! :)

I'm going to take a closer look at the impacts of the apple farm changes, and if that impacts existing farms. If it does, then we may have to use a development branch that continue alongside main, and gets merged into main when a release is ready. I'd prefer to make factory breaking changes only when they're really necessary.

I'll make a note in the title to not squash your commits so the info doesn't get lost! Please remind me in future PRs :)

**Comment by winex** [comment]
@bahrmichael as you may read in #184 - apple farms has been already occupying the 13-line by half (as being 12.5 tall), i've provided some screenshots there and my saves didn't show anything breaking, so i pushed that

one possible problem i see here: if anyone has wrongly-rounded (y+0.5) placement (ex: different CPU arch), so it could occupy the 14-nth line, but it should never happen as per Factorio's work on Lua :)))

**Comment by bahrmichael** [comment]
All good! Thanks for the explainer. If there's only a small chance that it breaks anything then we can put out a warning. I just don't want players to revisit all apple farms.

**Review by winex — COMMENTED** [review]
rebased against latest `v0.4.5` - looks good

**Comment by bahrmichael** [comment]
I did some tests yesterday. Fishery spawns and distribution look great!

I'm not sure what approach you're using regarding placement limits. I noticed for example that cities spawn close or on top of water, farms can spawn next to cliffs so that their water input is blocked, and something more. Will provide screenshots later!

**Comment by winex** [comment]
@bahrmichael oh, i've watched your stream yesterday (screens aren't needed that much). will add your feedback somewhere (probably in rbgen issue) - thanks. never tried `islands` map, btw -- i should check that

about cliffs and cities - yes, but it is using your code (cliff, water checks there), actually - didn't change this, yet. it seems some tweaks are needed

**Comment by bahrmichael** [comment]
> about cliffs and cities - yes, but it is using your code (cliff, water checks there), actually - didn't change this, yet. it seems some tweaks are needed

Gotcha! Then I'll take another look at the placement limitations when we prepare this for release :) That can be done last minute as we know enough about the code and when to restart map gen. Thanks for pushing this forward!

**Comment by bahrmichael** [comment]
In case you prefer screens over rewatching the stream, here are some notes:

Screenshot 2024-03-09 174828
Fisheries very close to each other, and on an island. Maybe not ideal as it's hard to access. I'm fine with rare cases of similar industries nearby each other, but in general it should have a good distribution.

Screenshot 2024-03-09 175027
City placed on water; may prevent the city's early expansion if nearby water or cliffs or resource patches.

Screenshot 2024-03-09 175245
This feels like there are a lot of factories close to each other in batches. Noticed this, but not sure if I'm for or against it. Just a different spawning behavior that we should think about.

Screenshot 2024-03-09 175300
Industry spawned very close to resource patch, which made me wonder if they could now spawn ON resource patches. I don't like the latter, but spawning close to patches is okay.

Screenshot 2024-03-09 175344
Industry spawned next to cliff. That blocks access to the industry, until the player has cliff explosives.

**Comment by winex** [comment]
this PR has **no** `rbgen` in it! it's just the base code (placeholders mostly) that it needs, i'm eager for this to be reviewed and merged as soon as possible. then i'll be preparing `[rbgen 2/2]: implementation` PR... -- and it will be as just few drop-in functions. wanted to make sure this doesn't break anything, so i split that into 2 parts

i'll post some answers in the correct discussion: #269!

**Comment by winex** [comment]
ok, i did some work against placement on resources - there were absolutely no code checking it.

while writing it i realised that it could cause performance penalty (if you have many resources added by mods), so a **new option** is going to be added (runtime, so no prob):
- `Skip generation check: resources` and similar new ones could follow this naming, if we ever need to skip smth else, like:
- `Skip generation check: some other`

this will go here, in the first half PR, because it doesn't depend on `rbgen`. an isolated commit will follow tomorrow..

_that **hard-difficulty** sandy map that has stuck cities + house offset bugs https://github.com/bahrmichael/factorio-tycoon/issues/89#issuecomment-1987308355 looks so cool being explored to 54MiB save - it has lots of sand areas, but when there is grass - `rbgen` just puts wheat and apple farms very nicely! i'm so happy. would be hard to provide a good screenshot - it's 8+ screens explored!_

**Comment by winex** [comment]
some more fixes, though it needs testing
+ new feature: population besides city names with runtime option to disable

**Comment by bahrmichael** [comment]
> it has lots of sand areas, but when there is grass - rbgen just puts wheat and apple farms very nicely! i'm so happy. would be hard to provide a good screenshot - it's 8+ screens explored!

That sounds really cool! It makes much more sense to have farms on grassland instead of deserts. Amazing work!

**Comment by bahrmichael** [comment]
I finally had time to review this in more detail! Amazing work :) I'll get this merged asap, and would rephrase the player-facing option for the resource skip. Thank you for implementing this!
